src/lib.rs

```diff
@@ -0,0 +1,534 @@
+        }
+    }
+
+    // TODO: Resize according to min set size instead of decrementing
+    pub(super) fn memory_resize_and_lock<'a>(
+        mut memory: &'a mut [usize],
+        allow_mem_resize: bool,
+    ) -> anyhow::Result<&'a mut [usize]> {
+        let page_size = unsafe {
+            let mut sysinfo: MaybeUninit<SYSTEM_INFO> = MaybeUninit::uninit();
+            GetNativeSystemInfo(sysinfo.as_mut_ptr());
+            let sysinfo = sysinfo.assume_init();
+            usize::try_from(sysinfo.dwPageSize).context("Failed to convert page size to usize")?
+        };
+
+        loop {
+            unsafe {
+                match VirtualLock(memory.as_mut_ptr().cast(), size_of_val(memory)) {
+                    Ok(()) => {
+                        info!("Successfully locked {}MB", size_of_val(memory));
+                        return Ok(memory);
+                    }
+                    Err(e) if e.code() == ERROR_WORKING_SET_QUOTA.to_hresult() => {
+                        if allow_mem_resize {
+                            match size_of_val(memory).checked_sub(page_size) {
+                                    Some(new_memsize) => {
+                                        warn!("Decremented memory size to {new_memsize}MB, retry memory locking");
+                                        memory = &mut memory[0..new_memsize / size_of::<usize>()];
+                                    }
+                                    None => bail!("Failed to lock any memory, memory size has been decremented to 0"),  
+                                }
+                        } else {
+                            bail!("VirtualLock failed to lock requested memory size: {e:?}")
+                        }
+                    }
+                    Err(e) => {
+                        return Err(e).context("VirtualLock failed to lock memory");
+                    }
+                }
+            }
+        }
+    }
+
+    pub(super) fn memory_unlock<'a>(memory: &'a mut [usize]) -> anyhow::Result<()> {
+        unsafe {
+            VirtualUnlock(memory.as_mut_ptr().cast(), size_of_val(memory))
+                .context("VirtualUnlock failed")
+        }
+    }
+}
+
+#[cfg(unix)]
+mod unix {
+    use {
+        crate::prelude::*,
+        libc::{__errno_location, mlock, munlock, sysconf, ENOMEM, _SC_PAGESIZE},
+        std::io::Error,
+    };
+
+    // TODO: Resize to RLIMIT_MEMLOCK instead of decrementing (note: memory might not be page aligned so locking the limit can still fail)
+    pub(super) fn memory_resize_and_lock(
+        mut memory: &mut [usize],
+        allow_mem_resize: bool,
+    ) -> anyhow::Result<&mut [usize]> {
+        let page_size = unsafe { sysconf(_SC_PAGESIZE) };
+        let page_size =
+            usize::try_from(page_size).context("Failed to convert page size to usize")?;
+
+        loop {
+            unsafe {
+                match mlock(memory.as_mut_ptr().cast(), size_of_val(memory)) {
```

Comment:
Same as above - `unsafe {}` only needs to be around the `mlock()` call. Also, instead of accessing `__errno_location()` you can rely on this: `Error::last_os_error().kind() == ErrorKind::OutOfMemory`. You can see a list of which Linux errors codes map to which `std::io::ErrorKind` variants [here](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/pal/unix/mod.rs).

Also, this code is getting a little bit arrow-y shaped. Don't be afraid to use early returns to avoid the [Arrow Anti-pattern](https://blog.codinghorror.com/flattening-arrow-code/). For example:

```Rust
loop {
    if unsafe { mlock(memory.as_mut_ptr().cast(), size_of_val(memory)) } == 0 {
        info!("Successfully locked {}MB", size_of_val(memory));
        return Ok(memory);
    }

    let e = Error::last_os_error();

    ensure!(e.kind() == ErrorKind::OutOfMemory, e.context("mlock failed"));
    ensure!(allow_mem_resize, e.context("mlock failed to lock requested memory size"));

    let new_len = memory
        .len()
        .checked_sub(usize_per_page)
        .context("failed to lock any memory, memory size has been decremented to 0")?;
    memory = &mut memory[0..new_len];
}
```